simulator/memory/argv_loader/argv_loader.cpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * argv_loader.cpp: loader of arguments to guest memory
+ * @author Rustem Yunusov <yunusov.rr@phystech.edu>
+ * Copyright 2019 MIPT-MIPS iLab project
+ */
+
+#include "argv_loader.h"
+#include <infra/exception.h>
+#include <cstring>
+
+ArgvLoader::ArgvLoader( const char* const* argv, const char* const* envp)
+        : argc( argv ? count_argc( argv) : 0)
+        , argv( argv)
+        , envp( envp)
+        , offset( 0)
+        , envp_offset( 0)
+{
+    if ( !argc)
+        throw InvalidArgs( "argc == 0");
+
+    if ( !argv)
+        throw InvalidArgs( "argv == nullptr");
+}
+
+size_t ArgvLoader::load_argv_to( const std::shared_ptr<FuncMemory>& mem, Addr addr)
+{
+    try { offset += mem -> memcpy_host_to_guest( addr + offset, byte_cast( &argc), 4); }
+    catch( FuncMemoryOutOfRange const& e) { throw ArgvLoaderError( std::string( "argc") + e.what()); }
```

Comment:
Catching exceptions do not work like that. Here you just change the type of exception, losing the backtrace information completely. The correct thing would be to remove `try/catch` block completely, so if an exception is thrown from `memcpy_host_to_guest`, it would unroll stack further and will be catched after that.